packages/@aws-cdk/aws-servicecatalog/lib/product-stack-asset-bucket.ts

```diff
@@ -0,0 +1,114 @@
+import { IBucket, BlockPublicAccess, Bucket, BucketEncryption } from '@aws-cdk/aws-s3';
+import { BucketDeployment, ISource, Source } from '@aws-cdk/aws-s3-deployment';
+import * as cdk from '@aws-cdk/core';
+import { Construct, IConstruct } from 'constructs';
+import { hashValues } from './private/util';
+/**
+ * Product stack asset bucket props.
+ */
+export interface ProductStackAssetBucketProps {
+  /**
+   * Name of s3 asset bucket deployed
+   *
+   * @default - None
+   */
+  readonly assetBucketName?: string;
+}
+
+/**
+ * A Service Catalog product stack asset bucket, which is similar in form to an Amazon S3 bucket.
+ * You can store multiple product stack assets and collectively deploy them to S3.
+ */
+export class ProductStackAssetBucket extends Construct {
+  private readonly bucketName: string;
+  private readonly bucket: IBucket;
+  private readonly assets: ISource[];
+
+  constructor(scope: Construct, id: string, props: ProductStackAssetBucketProps = {}) {
+    super(scope, id);
+
+    if (props.assetBucketName) {
+      this.bucketName = props.assetBucketName;
+    } else {
+      this.bucketName = this.generateBucketName(id);
+    }
+
+    this.bucket = new Bucket(scope, `${id}S3Bucket`, {
+      bucketName: this.bucketName,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+      encryption: BucketEncryption.KMS,
+      removalPolicy: cdk.RemovalPolicy.DESTROY,
+      autoDeleteObjects: true,
+    });
+
+    this.assets = [];
+
+    cdk.Aspects.of(this).add({
```

Comment:
what is the reasoning behind using `cdk.Aspects` to run deploy assets? not questioning, just wondering.

Reply:
It can be done without Aspects I believe, but Aspects fits the use case and allows us to cleanly implement our use case as well as abstract the additional code to the new `product-stack-asset-bucket` constuct rather than modifying and adding additional code to  `product-stack`.